### PR TITLE
Dfs solver

### PR DESCRIPTION
**Pull request by pbosler**
This PR adds the particle double Fourier series Poisson solver and a coupled particle/panel/grid solver for the barotropic vorticity equation.   However, some to-do items remain before we can merge:

- [x] Clean up `sphere_lpm_code` folder; all source should be in `lpm/src/dfs`
- [x] Create new test data folder, add its location to `LpmConfig.h.in`
- [x] Clean up test data files: delete the unnecessary ones and move the rest to the test data folder.

### CODE REVIEW

File: sphere_lpm_code/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,103 @@
+cmake_minimum_required(VERSION 3.12)
+
```
Comment (pbosler) [inline sphere_lpm_code/CMakeLists.txt:2]: I don't think we need this directory in the main repo.   @michaelchiwere1 please confirm and delete from the branch, if we can.

File: sphere_lpm_code/datafiles/coeffs_imag.bin
Comment (pbosler) [inline sphere_lpm_code/datafiles/coeffs_imag.bin:1]: If these binary files are data for unit testing, let's create a new folder `lpm/tests/data` and put them in it.

Let's delete any that are not necessary.

**Review by pbosler — COMMENTED** [review]
We need to do some cleanup before we merge this branch.  Mainly, we need to delete the `sphere_lpm_code` folder (after making sure that everything we need from it is already in `lpm/src/dfs`).    We need to either delete or move the binary files to a new test data folder.

File: sphere_lpm_code/CMakeLists.txt (hunk 1 above)
Comment (michaelchiwere1) [reply sphere_lpm_code/CMakeLists.txt:2]: I think we should not need it, indeed.

**Comment by pbosler** [comment]
This PR has some infrastructure changes inside of it, for example, the TPL build scripts and the changes to gather/scatter mesh.   I'm going to pull those out into a separate pull request since they're not just applicable to DFS.

**Comment by pbosler** [comment]
I'm going to merge this, so we can make feature branches (like for RK3 and RK4) easier.   I'll make separate issues for the things we still need to do to clean this up.

**Pull request merged**